### DISCUSSION TITLE
Use of pseudo-design matrices to avoid label switching

### DISCUSSION THREAD

**Discussion by DEHewitt** [bmcclintock/momentuHMM#57] [Q&A] [answered]
Hi,

I am currently preparing data for an analysis using `momentuHMM`. My data is temporally-irregular and has some measurement error and I would like to use the multiple imputation approach (as I have had difficulties with fitting state-switching SSMs to this data). I've read the various other issues (e.g. #42) that relate to the use of pseduo-design matrices to avoid label switching and I _think_ I'm starting to get it. I note that it is trivial and I'm fairly certain I've had the "a-ha!" moment you describe in #42, but I just want to be sure. 

It is my understanding that for a 3-state HMM which requires zero-inflation the following would work: 
```
nbStates <- 3
dist <- list(step = "gamma", angle = "wrpcauchy")
stepDM <- matrix(c(1,0,0,0,0,0,0,0,0,
                   0,1,0,0,0,0,0,0,0,
                   0,0,1,0,0,0,0,0,0,
                   0,0,0,1,0,0,0,0,0,
                   0,0,0,0,1,0,0,0,0,
                   0,0,0,0,0,1,0,0,0,
                   0,0,0,0,0,0,1,0,0,
                   0,0,0,0,0,0,0,1,0,
                   0,0,0,0,0,0,0,0,1),
                 nrow = 3*nbStates,
                 byrow = T,
                 dimnames = list(c(paste0("mu_", 1:nbStates),
                                   paste0("sd_", 1:nbStates),
                                   paste0("zeromass_", 1:nbStates)),
                                 c(paste0("mu_", 1:nbStates),
                                   paste0("sd_", 1:nbStates),
                                   paste0("zeromass_", 1:nbStates))))
```

Then, if I wish to impose a constraint such that `mu_3` > `mu_1` I would change the `0` to a `1` in the corresponding cell (i.e. the third row in the first column) and specify the bounds using an upper bound for `mu_1` that is less than the upper bound for `mu_3`, e.g.:
```
stepuserBounds <- matrix(c(0,0,0,0,0,0,0,0,0, 
                           2,2,4,2,2,4,0.1,0.2,0.3), # numbers chosen arbitrarily for illustration
                         9,2,
                         dimnames=list(colnames(stepDM),c("lower","upper")))
```

All of which can then be passed to `getParDM()` to convert the initial parameters from the natural scale to the working scale.

Have I had the "a-ha!" moment or do I have more reading to do? Any advice you can provide would be extremely appreciated!

Thanks,
Dan

**Comment by bmcclintock** [comment, marked as answer]
You're halfway there. You are correct that in order to constrain `mu_3` > `mu_1`, `stepDM` would be:
```
stepDM <- matrix(c(1,0,0,0,0,0,0,0,0,
                   0,1,0,0,0,0,0,0,0,
                   1,0,1,0,0,0,0,0,0,
                   0,0,0,1,0,0,0,0,0,
                   0,0,0,0,1,0,0,0,0,
                   0,0,0,0,0,1,0,0,0,
                   0,0,0,0,0,0,1,0,0,
                   0,0,0,0,0,0,0,1,0,
                   0,0,0,0,0,0,0,0,1),
                 nrow = 3*nbStates,
                 byrow = T,
                 dimnames = list(c(paste0("mu_", 1:nbStates),
                                   paste0("sd_", 1:nbStates),
                                   paste0("zeromass_", 1:nbStates)),
                                 c("wmu_13","wmu_2","wmu_3",
                                     paste0("wsd_", 1:nbStates),
                                     paste0("wzeromass_", 1:nbStates))))
```
However, it is not `userBounds` that you want to use. `userBounds` puts *known* constraints on the real scale parameters (e.g. 0 < `mu_1` < 2, 0 < `mu_2` < Inf, 2 < `mu_3` < Inf), but these values are not typically known. Instead you need to use `workBounds` to put the constraints on the working scale parameters (e.g. `wmu_13`, `wmu_2`, and `wmu_3`):
```
stepworkBounds <- matrix(c(-Inf,Inf,
                           -Inf,Inf,
                              0,Inf,
                           -Inf,Inf,
                           -Inf,Inf,
                           -Inf,Inf,
                           -Inf,Inf,
                           -Inf,Inf,
                           -Inf,Inf),nrow=ncol(stepDM),byrow = TRUE,dimnames=list(c("wmu_13","wmu_2","wmu_3",
                                                                                    paste0("wsd_", 1:nbStates),
                                                                                    paste0("wzeromass_", 1:nbStates)),
                                                                                  c("lower","upper")))
```
Thus `mu_1 = exp(1*wmu_13+0*wmu_2+0*wmu_3+0*wsd_1+...+0*wzeromass_3)`, `mu_2 = exp(0*wmu_13+1*wmu_2+0*wmu_3+0*wsd_1+...+0*wzeromass_3)` and `mu_3 = exp(1*wmu_13+0*wmu_2+1*wmu_3+0*wsd_1+...+0*wzeromass_3)`, and `mu_3` must be greater than `mu_1` because `wmu_3` must be positive (i.e., between 0 and Inf).

**Reply by DEHewitt** [reply]
Awesome, thanks for the quick response. Also, thanks for developing such a cool package (and thorough vignette). I'm super excited to see the results of this project! 

I was getting a little confused by `userBounds` and `workBounds` - I'll have to go over the vignette again. As far as avoiding label switching: the diagonal `1`'s through the `DM` are what does the trick then, effectively turning parameters "on" (or off with `0`'s) when they're being estimated? I ask as I don't want to include any constraints on the turning angles of each state, but assume I still need a pseudo-`DM` to avoid label switching if I'm using multiple imputation.
